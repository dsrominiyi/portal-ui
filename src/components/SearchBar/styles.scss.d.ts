export const searchContainer: string;
export const resultsContainer: string;
export const searchInput: string;
export const searchIcon: string;
export const results: string;
export const dropdownItem: string;
export const matchedText: string;
