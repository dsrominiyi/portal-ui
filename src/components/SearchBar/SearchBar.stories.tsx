import React from 'react';
import { storiesOf } from '@storybook/react';

import SearchBar from '.';

const stories = storiesOf('SearchBar', module);

const render = (
  hits?: string[],
  buildUrl?: (result: SearchHit) => string,
  buildCustomerResultsUrl?: (term: string) => string
) => {
  const searchResults = {
    found: hits ? hits.length : 0,
    start: 0,
    hit: hits
      ? hits.map((hit: string, index: number) => ({
          id: String(index),
          fields: { id: String(index), text: hit }
        }))
      : ([] as SearchHit[])
  };

  return () => (
    <SearchBar
      searchResults={searchResults}
      searchFetchResults={() => ({
        type: '',
        payload: undefined
      })}
      buildUrl={buildUrl}
      buildMoreResultsUrl={buildCustomerResultsUrl}
    />
  );
};

const buildUrlFunc = (result: SearchHit) => `/customer/${result.fields.id}`;
const moreResultsFunc = (term: string) => `/customer/search/${term}`;

export const storyRenders = {
  noResults: render([]),
  twoResults: render(['Toby Test', 'Tim Test']),
  sixResults: render(['Toby Test', 'Tim Test', 'Tom Test', 'Ted Test', 'Tab Test', 'Tod Test']),
  eightteenResults: render([
    'Toby Test',
    'Tim Test',
    'Tom Test',
    'Ted Test',
    'Tab Test',
    'Tod Test',
    'Toby Test',
    'Tim Test',
    'Tom Test',
    'Ted Test',
    'Tab Test',
    'Tod Test',
    'Toby Test',
    'Tim Test',
    'Tom Test',
    'Ted Test',
    'Tab Test',
    'Tod Test'
  ]),
  withbuildUrlAndBuildResultsUrlUnderTen: render(
    ['Toby Test', 'Tim Test'],
    buildUrlFunc,
    moreResultsFunc
  ),
  withbuildUrlAndBuildResultsUrlOverTen: render(
    [
      'Toby Test',
      'Tim Test',
      'Tom Test',
      'Ted Test',
      'Tab Test',
      'Tod Test',
      'Toby Test',
      'Tim Test',
      'Tom Test',
      'Ted Test',
      'Tab Test',
      'Tod Test',
      'Toby Test',
      'Tim Test',
      'Tom Test',
      'Ted Test',
      'Tab Test',
      'Tod Test'
    ],
    buildUrlFunc,
    moreResultsFunc
  )
};

stories.add('with no hits', storyRenders.noResults);
stories.add('with 2 hits', storyRenders.twoResults);
stories.add('with 6 hits', storyRenders.sixResults);
stories.add('with 18 hits', storyRenders.eightteenResults);
