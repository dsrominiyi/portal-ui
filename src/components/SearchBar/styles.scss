@import '../../style/_vars';

.search-container {
  width: 100%;
  display: flex;
  flex: 1;
  flex-direction: column;
  justify-content: center;
}

.results-container {
  position: relative;
}

.search-input {
  flex: 1;
}

.search-icon {
  svg {
    > * {
      stroke: $brand-charcoal;
    }
  }
}

.results {
  position: absolute;
  max-height: 60vh;
  min-width: 10rem;
  padding: 0.5rem 0;
  margin: 0.125rem 0 0;
  outline: none;
  border: 1px solid #e3e6f0;
  border-radius: 0.35rem;
  font-size: 1rem;
  color: $brand-heavy-thunder;
  background-color: $brand-white;
  text-align: left;
  list-style: none;
  z-index: 50;
  box-shadow: 0 0.15rem 1.75rem 0 rgba(58, 59, 69, 0.15);
  cursor: pointer;

  .dropdown-item {
    background-color: $brand-white;
    display: block;
    padding: 0.25rem 1.5rem;
    font-weight: 400;
    white-space: nowrap;
    text-decoration: none;

    &:hover {
      background-color: $brand-dark-white;
    }

    &:link,
    &:visited {
      color: $brand-heavy-thunder;
    }

    &:active,
    &:link:active {
      color: $brand-white;
      background-color: $brand-success;
    }

    .matched-text {
      font-weight: 600;
    }
  }
}
