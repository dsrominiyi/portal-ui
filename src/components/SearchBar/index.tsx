import React, { Component, ReactNode, Fragment, MouseEvent } from 'react';
import SearchIcon from 'uikit-icons/lib/Search';
import TextField from '../TextField';

import styles from './styles.scss';
import Link from '../Link';

import debounce from '../../helpers/debounce';

export interface Props {
  additionalClasses?: string[];
  label?: string;
  searchResults: SearchResults;
  attributes?: { [attr: string]: string };
  searchFetchResults(
    searchTerm: string,
    questionId?: string
  ): { type: string; payload?: Promise<SearchResponse> };
  buildResultText?(fields: SearchHit['fields']): string;
  onClick?(event: MouseEvent<Element>): void;
  buildUrl?(result: SearchHit): string;
  buildMoreResultsUrl?(term: string): string;
}

interface State {
  searchTerm: string;
  fieldFocused: boolean;
}

export class SearchBar extends Component<Props> {
  public static isSearchSuggestion = (result: SearchResult): result is SearchSuggestion => {
    if ((result as SearchSuggestion).suggestion) {
      return true;
    }
    return false;
  };

  public static isSearchSuggestionList = (
    results: SearchResults
  ): results is SearchSuggestionList => {
    if (results && (results as SearchSuggestionList).suggestions) {
      return true;
    }
    return false;
  };

  public state: State = {
    searchTerm: '',
    fieldFocused: false
  };

  private onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { searchFetchResults } = this.props;
    this.setState({ searchTerm: event.target.value });
    debounce(searchFetchResults, 200, false)(event.target.value);
  };

  private buildResults = () => {
    const { fieldFocused } = this.state;
    const { searchResults, onClick, buildUrl, buildMoreResultsUrl } = this.props;
    let resultSet: SearchResult[] = [];
    let renderedResults: ReactNode[] = [];

    if (searchResults) {
      resultSet = SearchBar.isSearchSuggestionList(searchResults)
        ? searchResults.suggestions
        : searchResults.hit;

      if (resultSet) {
        renderedResults = resultSet
          .filter((_value: SearchResult, index: number) => index < 10)
          .map((result: SearchResult) => (
            <li key={result.id}>
              <Link
                classNames={[styles.dropdownItem]}
                label={this.getResultText(result)}
                href={buildUrl ? buildUrl(result as SearchHit) : undefined}
                onClick={onClick}
              />
            </li>
          ));
      }
      if (searchResults.found > 10) {
        renderedResults.push(
          <li key="more-results">
            <Link
              classNames={[styles.dropdownItem]}
              label={`+ ${searchResults.found - 10} more results - click to view all`}
              href={buildMoreResultsUrl && buildMoreResultsUrl(this.state.searchTerm)}
              onClick={onClick}
            />
          </li>
        );
      }

      return (
        fieldFocused &&
        !!renderedResults.length && <ul className={styles.results}>{renderedResults}</ul>
      );
    }
  };

  private getResultText = (result: SearchResult): ReactNode => {
    const { buildResultText } = this.props;
    let resultText: string;

    if (SearchBar.isSearchSuggestion(result)) {
      resultText = result.suggestion;
    } else if (buildResultText) {
      resultText = buildResultText(result.fields);
    } else {
      const resultValues = Object.keys(result.fields).map(fieldKey => result.fields[fieldKey]);
      resultText = resultValues.join(' - ');
    }

    return this.highlightMatch(resultText);
  };

  private highlightMatch = (resultText: string): ReactNode => {
    const { searchTerm } = this.state;
    const matchIndex = resultText.toLowerCase().indexOf(searchTerm.toLowerCase());
    const postMatchIndex = matchIndex + searchTerm.length;

    return !searchTerm || matchIndex === -1 ? (
      resultText
    ) : (
      <Fragment>
        {resultText.slice(0, matchIndex)}
        <strong className={styles.matchedText}>
          {resultText.slice(matchIndex, postMatchIndex)}
        </strong>
        {resultText.slice(postMatchIndex)}
      </Fragment>
    );
  };

  public render() {
    const { additionalClasses = [], label, attributes } = this.props;
    const { searchTerm } = this.state;

    return (
      <div className={[styles.searchContainer, ...additionalClasses].join(' ')}>
        <TextField
          additionalClasses={[styles.searchInput]}
          value={searchTerm}
          label={label}
          placeholder="Search..."
          icon={<SearchIcon className={styles.searchIcon} />}
          reverseIconPosition
          onChange={this.onChange}
          onFocus={() => this.setState({ fieldFocused: true })}
          onBlur={() => this.setState({ fieldFocused: false })}
          {...attributes}
        />
        <div className={styles.resultsContainer}>{this.buildResults()}</div>
      </div>
    );
  }
}

export default SearchBar;
