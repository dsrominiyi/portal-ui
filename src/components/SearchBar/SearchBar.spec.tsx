import { shallow, ShallowWrapper } from 'enzyme';
import { storyRenders } from './SearchBar.stories';
import testStorySnapshots from '../../helpers/test/testStorySnapshots';

import SearchBar from '.';

let component: ShallowWrapper;
const hitResults = {
  start: 0,
  found: 3,
  hit: [
    { id: '001', fields: { description: 'test 1', netval: '001' } },
    { id: '002', fields: { description: 'test 2', netval: '002' } },
    { id: '003', fields: { description: 'test 3', netval: '003' } }
  ]
};

describe('User', () => {
  beforeEach(() => {
    component = shallow(storyRenders.twoResults());
  });

  test('component story renders match snapshots', () => {
    testStorySnapshots(storyRenders);
  });

  test('isSearchSuggestion returns true if results are search suggestions type', () => {
    const searchResults = {
      id: '001',
      score: 3,
      suggestion: 'string'
    };

    expect(SearchBar.isSearchSuggestion(searchResults)).toBe(true);
  });

  test('isSearchSuggestion returns false if results are search hits type', () => {
    const searchResults = {
      id: '001',
      fields: {}
    };

    expect(SearchBar.isSearchSuggestion(searchResults)).toBe(false);
  });

  test('SearchBar updates phrase as phrase is typed', () => {
    component = shallow(storyRenders.noResults());
    const searchTerm = 'Test Search';
    const searchTermInput = component.find('TextField');
    const event = { target: { value: searchTerm } };
    searchTermInput.simulate('change', event);

    expect(component.state('searchTerm')).toBe(searchTerm);
  });

  test('SearchBar highlights matches for suggestions', () => {
    const searchTerm = 'Tes';
    component.setState({ searchTerm });
    const firstResult = component.find('Link').first();
    const matchedTextElement = (firstResult.props().label as any).props.children[1];

    expect(matchedTextElement.props.children).toBe(searchTerm);
  });

  test('SearchBar highlights matches for hits', () => {
    const searchTerm = 'Tes';
    component.setState({ searchTerm });
    component.setProps({ searchResults: hitResults });
    const firstResult = component.find('Link').first();
    const matchedTextElement = (firstResult.props().label as any).props.children[1];

    expect(matchedTextElement.props.children).toBe(searchTerm.toLowerCase());
  });

  test('SearchBar highlights matches for hits where showResultFields is defined', () => {
    const searchTerm = 'Tes';
    component.setState({ searchTerm });
    component.setProps({ searchResults: hitResults, showResultFields: ['description'] });
    const firstResult = component.find('Link').first();
    const matchedTextElement = (firstResult.props().label as any).props.children[1];

    expect(matchedTextElement.props.children).toBe(searchTerm.toLowerCase());
  });

  test('SearchBar does not highlight when a returned phrase does not match the typed phrase', () => {
    const searchTerm = 'tim';
    component.setState({ searchTerm });
    const firstResult = component.find('Link').first();

    expect(firstResult.props().label).toBe('Toby Test');
  });

  test('SearchBar does not render results if there is no searchTerm', () => {
    const searchTerm = 'Test';
    component.setState({ searchTerm });
    component.setProps({ searchResults: hitResults });
    component.setState({ searchTerm: '' });

    expect(component.find('.results').length).toBe(0);
  });
});
