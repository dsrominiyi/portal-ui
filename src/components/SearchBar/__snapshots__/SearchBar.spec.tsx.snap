// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User component story renders match snapshots: eightteenResults 1`] = `
<div
  className="searchContainer"
>
  <TextField
    additionalClasses={
      Array [
        "searchInput",
      ]
    }
    icon={
      <SvgSearch
        className="searchIcon"
      />
    }
    onChange={[Function]}
    placeholder="Search..."
    reverseIconPosition={true}
    value=""
  />
  <div
    className="resultsContainer"
  />
</div>
`;

exports[`User component story renders match snapshots: noResults 1`] = `
<div
  className="searchContainer"
>
  <TextField
    additionalClasses={
      Array [
        "searchInput",
      ]
    }
    icon={
      <SvgSearch
        className="searchIcon"
      />
    }
    onChange={[Function]}
    placeholder="Search..."
    reverseIconPosition={true}
    value=""
  />
  <div
    className="resultsContainer"
  />
</div>
`;

exports[`User component story renders match snapshots: sixResults 1`] = `
<div
  className="searchContainer"
>
  <TextField
    additionalClasses={
      Array [
        "searchInput",
      ]
    }
    icon={
      <SvgSearch
        className="searchIcon"
      />
    }
    onChange={[Function]}
    placeholder="Search..."
    reverseIconPosition={true}
    value=""
  />
  <div
    className="resultsContainer"
  />
</div>
`;

exports[`User component story renders match snapshots: twoResults 1`] = `
<div
  className="searchContainer"
>
  <TextField
    additionalClasses={
      Array [
        "searchInput",
      ]
    }
    icon={
      <SvgSearch
        className="searchIcon"
      />
    }
    onChange={[Function]}
    placeholder="Search..."
    reverseIconPosition={true}
    value=""
  />
  <div
    className="resultsContainer"
  />
</div>
`;

exports[`User component story renders match snapshots: withbuildUrlAndBuildResultsUrlOverTen 1`] = `
<div
  className="searchContainer"
>
  <TextField
    additionalClasses={
      Array [
        "searchInput",
      ]
    }
    icon={
      <SvgSearch
        className="searchIcon"
      />
    }
    onChange={[Function]}
    placeholder="Search..."
    reverseIconPosition={true}
    value=""
  />
  <div
    className="resultsContainer"
  />
</div>
`;

exports[`User component story renders match snapshots: withbuildUrlAndBuildResultsUrlUnderTen 1`] = `
<div
  className="searchContainer"
>
  <TextField
    additionalClasses={
      Array [
        "searchInput",
      ]
    }
    icon={
      <SvgSearch
        className="searchIcon"
      />
    }
    onChange={[Function]}
    placeholder="Search..."
    reverseIconPosition={true}
    value=""
  />
  <div
    className="resultsContainer"
  />
</div>
`;
